pkg/yum/repository.go

```diff
@@ -0,0 +1,299 @@
+package yum
+
+import (
+	"compress/gzip"
+	"encoding/xml"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"path"
+)
+
+// Package metadata of a given package
+type Package struct {
+	Type     string   `xml:"type,attr"`
+	Name     string   `xml:"name"`
+	Arch     string   `xml:"arch"`
+	Version  Version  `xml:"version"`
+	Checksum Checksum `xml:"checksum"`
+	Summary  string   `xml:"summary"`
+}
+
+type Version struct {
+	Version string `xml:"ver,attr"`
+	Release string `xml:"rel,attr"`
+	Epoch   int32  `xml:"epoch,attr"`
+}
+
+type Checksum struct {
+	Value string `xml:",chardata"`
+	Type  string `xml:"type,attr"`
+}
+
+// Repomd metadata of the repomd of a repository
+type Repomd struct {
+	XMLName      xml.Name `xml:"repomd"`
+	Data         []Data   `xml:"data"`
+	Revision     string   `xml:"revision"`
+	RepomdString *string  `xml:"-"`
+}
+
+type Data struct {
+	Type     string   `xml:"type,attr"`
+	Location Location `xml:"location"`
+}
+type Location struct {
+	Href string `xml:"href,attr"`
+}
+
+type YummySettings struct {
+	Client *http.Client
+	URL    *string
+}
+
+type YumRepository interface {
+	Configure(settings YummySettings)
+	Packages() (packages []Package, statusCode int, err error)
+	Repomd() (repomd *Repomd, statusCode int, err error)
+	Signature() (repomdSignature *string, statusCode int, err error)
+}
+
+type Repository struct {
+	settings        YummySettings
+	packages        []Package // Packages repository contains
+	repomdSignature *string   // Signature of the repository
+	repomd          *Repomd   // Repomd of the repository
+}
+
+func NewRepository(settings YummySettings) (Repository, error) {
+	if settings.Client == nil {
+		return Repository{}, fmt.Errorf("client cannot be nil")
+	}
+	if settings.URL == nil {
+		return Repository{}, fmt.Errorf("url cannot be nil")
+	}
+	return Repository{settings: settings}, nil
+}
+
+func (r *Repository) Configure(settings YummySettings) {
+	if settings.Client != nil {
+		r.settings.Client = settings.Client
+	}
+	if settings.URL != nil {
+		r.settings.URL = settings.URL
+	}
+}
+
+// Repomd populates r.Repomd with repository's repomd.xml metadata. Returns Repomd, response code, and error.
+// If the repomd was successfully fetched previously, will return repomd.
+func (r *Repository) Repomd() (*Repomd, int, error) {
+	var result Repomd
+	var err error
+	var resp *http.Response
+	var repomdURL string
+
+	if r.repomd != nil {
+		return r.repomd, 0, nil
+	}
```

Comment:
I also considered having a store on the `Repository` structure that gets the previously fetched data, the benefit of that being it could also save the HTTPCode/Error.

```go
type repomdStore struct {
    Repomd Repomd
    Code int
    Error error
}
```

Reply:
I think this is fine, it will only get cached if its successful, further attempts will re-try which i think is a good thing